File: app/src/main/java/com/waveneuro/domain/usecase/login/ConfirmTokenUseCase.java
package com.waveneuro.domain.usecase.login;

import com.asif.abase.domain.base.ObservableUseCase;
import com.asif.abase.domain.base.UseCaseCallback;
import com.waveneuro.data.DataManager;
import com.waveneuro.data.model.request.login.ConfirmTokenRequest;
import com.waveneuro.data.model.response.login.ConfirmTokenResponse;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Observable;

public class ConfirmTokenUseCase extends ObservableUseCase<ConfirmTokenResponse> {

    private final DataManager dataManager;

    private ConfirmTokenRequest request;

    @Inject
    public ConfirmTokenUseCase(DataManager dataManager) {
        this.dataManager = dataManager;
    }

    @Override
    public Observable<ConfirmTokenResponse> buildUseCaseSingle() {
        return dataManager.confirmToken(request);
    }

    public void execute(ConfirmTokenRequest request, UseCaseCallback<ConfirmTokenResponse> useCaseCallback) {
        this.request=request;
        super.execute(useCaseCallback);
    }

}

File: app/src/main/java/com/waveneuro/domain/usecase/login/ConfirmTokenUseCase.kt
package com.waveneuro.domain.usecase.login

import com.waveneuro.data.api.user.UserApi
import com.waveneuro.data.api.user.model.mfa.ApiConfirmTokenRq
import com.waveneuro.domain.model.mfa.ConfirmTokenRs
import com.waveneuro.domain.model.mfa.MfaMapperImpl
import javax.inject.Inject

class ConfirmTokenUseCase @Inject constructor(
    private val userApi: UserApi,
    private val mapper: MfaMapperImpl
) {

    suspend fun confirmToken(username: String, mfaCode: String, session: String): ConfirmTokenRs =
        mapper.fromApiToDomain(
            userApi.confirmSoftwareToken(ApiConfirmTokenRq(username, mfaCode, session))
        )

}